Add new option to write and load new-relic.js file

GCheung55 opened the pull request:
Fix for #6.

This adds options to write and load an external new-relic.js file.

Commit:
Add new option to write new relic js file and insert script tag to load js file externally via contentFor.

GCheung55 commented:
@sir-dunxalot I'm not sure what's causing the build to fail with `network read ECONNRESET` but I've read that a rebuild could fix the issue?

I'm also not sure about how to write tests for this.

Commit:
Adding different test scenarios to travis.

Commit:
Only write script tag to load externa lfile when loadExternal and outputPath.

Commit:
Add outputPath to test scenario 2.

Commit:
Add a 3rd case to include throwing an error if outputPath is undefined but loading an external file is expected.

Commit:
Removing an case 3 because it was used for testing expected failure.
Update to load new-relic.js file into vendor.js file by default.
Optionally output new-relic.js file to public.

sir-dunxalot commented:
I've restarted the build just incase it was a one-off issue. As we discussed in Slack, I'm a little hesitant to marge this if we can't show the test suite is passing.

A good test could use a Broccoli/node reader to check the file exists in the `tmp` tree. I have another addon, `ember-cli-concat`, which runs similar tests so I reccomend having a look at how [those tests are set up](https://github.com/sir-dunxalot/ember-cli-concat/blob/master/tests/acceptance/concatenation-test.js). Also, `ember-cli-modernizr` has tests in this realm [here](https://github.com/sir-dunxalot/ember-cli-modernizr/blob/master/tests/helpers/should/build-with.js#L22).

GCheung55 commented:
Right! I wouldn't merge a PR if the tests weren't passing either. It still looks like npm isn't behaving even from your restart of the build.

I didn't know that broccoli could be loaded in the browser. I'll look at both of the repos as inspiration.

Commit:
Test writeTrackingCodeTree actually writes a file and the contents are as expected.
Exclude node_tests/tmp dir that broccoli writes to in the test.

Commit:
Missed a dependency.

GCheung55 commented:
@sir-dunxalot Passing (finally)! Please review.

sir-dunxalot commented on ember-cli-build.js:
```diff
@@ -1,18 +1,32 @@
+module.exports = function(defaults) {
+  var options;
+
+  switch (process.env.TEST_SCENARIO) {
```
Does this affect builds in consuming apps? I wonder if this code is best set in `.travis.yml` or something similar. For example, using a matrix for Travis tests: https://docs.travis-ci.com/user/environment-variables/

GCheung55 replied:
`.travis.yml` was modified to include a matrix for 2 scenarios. According to https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/ember-addon.js#L14, `ember-cli/lib/broccoli/ember-addon` is only used in development.

sir-dunxalot commented on index.js:
```diff
@@ -1,13 +1,46 @@
+  importToVendor: true,
+
+  included: function(app) {
+    this._super.included ? this._super.included.apply(this, arguments) : this._super(app);
```
Could you add a code comment explaining why this is necessary?

sir-dunxalot commented on index.js:
```diff
@@ -1,13 +1,46 @@
+    this._super.included ? this._super.included.apply(this, arguments) : this._super(app);
+
+    var env  = process.env.EMBER_ENV;
+    var options = (this.app && this.app.options && this.app.options['ember-new-relic']) || {};
```
Don't this return a boolean inside the parentheses?

GCheung55 replied:
It will return the last truthy value, so it will return the value of `this.app.options['ember-new-relic']`.

An example of this would be:

``` javascript
var options = {};

console.log((options && options.foo) || 'nope');

options.foo = {baz: true};
console.log((options && options.foo) || 'nope');
```

sir-dunxalot commented on index.js:
```diff
@@ -1,13 +1,46 @@
+    }
+  },
+
+  isDevelopingAddon: function() {
```
Should this function still be here?

GCheung55 replied:
No, I will remove it since it's for development only.

sir-dunxalot commented on node_tests/index-test.js:
```diff
@@ -1,14 +1,17 @@
 QUnit.module("ember-new-relic | When config['ember-new-relic'].spaMonitoring is false", {
   setup: function() {
-    this.newRelicConfig = {
+    this.newRelicConfig = EmberNewRelic.getNewRelicConfig({
```
Where is `EmberNewRelic` defined in this doc?

GCheung55 replied:
I followed the example of other code in this file and it looks like `EmberNewRelic` was being defined by `node_tests/run.js` by `qunit`.

sir-dunxalot reviewed (commented):
I've left a few questions. Once those are addressed I'll pull this down and check it out locally. Overall this looks like a good step forwards for the library!

sir-dunxalot commented:
Comments looks good. Let me know when you've made the necessary changes and we can merge :)

Commit:
Call the parent class' `included` method instead of through `_super`.
`isValidNewRelicConfig` property for determining if newRelicConfig has applicationId and licenseKey
Write warning line if the applicationId or licenseKey properties are missing - shouldn't block.
Update rest of code to check against `isValidNewRelicConfig` property.
Update node tests with `isValidNewRelicConfig` property.

GCheung55 commented:
@sir-dunxalot I've left a comment in my commit about printing out a warning. Please do a second pass. :)

One one last thing... documentation. Do you want me to do update the README.md?

Commit:
Instead of using `this.ui.writeLine` for all versions of ember-cli, use `writeWarnLine` if it exists, but fallback to writeLine.
This allows for color coded messages that newer versions of ember-cli provides.

sir-dunxalot commented:
Alright, these changes all look good. Sorry - I missed your question above. Yes, documentation would be great. That will also help me fully understand the implementation :)

As soon as that is done I will merge and release! We could release later tonight or tomorrow.

Commit:
Update README with instructions for new changes.
